text/0000-app-mixins.md

```diff
@@ -0,0 +1,214 @@
+cache = false
+```
+
+Where:
+
+* `paths` = a list of paths to exclude from the layer
+* `cache` = if true, the paths will be excluded from the launch image layer, but will be included in the cache layer.
+
+## Rebasing an App
+
+App that uses stackpacks can be rebased as normal. Stackpacks are expected to retain ABI compatibility.
+
+## Example: Apt Buildpack
+
+ A buildpack that can install an arbitrary list of mixins would have a `buildpack.toml` like this:
+
+ ```toml
+[buildpack]
+id = "example/apt"
+privileged = true
+
+[[stacks]]
+id = "io.buildpacks.stacks.bionic"
+```
+
+Its `bin/detect` would have the following contents:
+
+ ```bash
+#!/usr/bin/env bash
+
+cat <<TOML >"$2"
+[[mixins]]
+name = "[^=]+"
+TOML
+```
+
+Its `bin/build` would have the following contents:
+
+ ```bash
+#!/usr/bin/env bash
+
+apt update -y
+
+for package in $(cat $3 | yj -t | jq -r ".entries | .[] | .name"); do
+  apt install $package
+done
+
+cat << EOF > launch.toml
+[[excludes]]
+paths = [ "/var/cache" ]
+cache = true
+EOF
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- Mixins are less flexible that a generic [root buildpack](https://github.com/buildpacks/rfcs/blob/app-image-ext-buildpacks/text/0000-app-image-extensions.md).
+
+# Alternatives
+[alternatives]: #alternatives
+
+- [Add root buildpack as interface for app image extension RFC](https://github.com/buildpacks/rfcs/pull/77)
+- [App Image Extensions (OS packages)](https://github.com/buildpacks/rfcs/pull/23)
+
+# Prior Art
+[prior-art]: #prior-art
+
+- [RFC-0006: Stage specific mixins](https://github.com/buildpacks/rfcs/blob/main/text/0006-stage-specific-mixins.md)
+- The term "non-idempotent" is used in section 9.1.2 of [Hypertext Transfer Protocol -- HTTP/1.1 (RFC 2616)](https://www.w3.org/Protocols/rfc2616/rfc2616-sec9.html)
+
+# Unresolved Questions
+[unresolved-questions]: #unresolved-questions
+
+- How should a buildpack be identified as a stackpack?
+    - Not everyone likes `privileged`
+    - a special `key` in the `buildpack.toml`?
+    - a `stackpack.toml`?
+- Should the stackpack's detect have read-only access to the app?
+    - In this proposal it does, and in the future a more generic "root buildpack" construct would need it too.
+
+# Spec. Changes (OPTIONAL)
+[spec-changes]: #spec-changes
+
+
+## Stackpacks
+
+Stackpacks are identical to other buildpacks, with the following exceptions:
+
+1. The `<layers>` directory is NOT writable
+1. The working directory WILL NOT contain application source code.
+1. All changes made to the filesystem (with the exception of `/tmp`) during the execution of the stackpack's `bin/build` will be snaphotted and stored as a single layer.
+1. A `launch.toml` WILL NOT be honored.
+
+## launch.toml (TOML)
+
+```
+[[excludes]]
+paths = ["<sub-path glob>"]
+cache = false
+```
+
+Where:
+
+* `paths` = a list of paths to exclude from the layer
+* `cache` = if true, the paths will be excluded from the launch image layer, but will be included in the cache layer.
```

Comment:
Doesn't this conflict with line 175

> A `launch.toml` WILL NOT be honored.

Sorry, I know this is still a draft, just trying to get a sense of it.

Reply:
Addressed in 9f39572